### ISSUE TITLE
Consumer does not shutdown when running in a different thread

### ISSUE THREAD

**Issue by deepchatterjeeligo** [astronomy-commons/genesis-client#57]
While the example below uses the hop-client to read the messages. I have tracked down the issue to the consumer in adc-streaming.

This issue is relevant when one is running a listener on a different thread, listening to some topic, where we want to set a sentinel variable `running` to `False` in the main thread for the listener to stop. Issue: If there is no message in the topic, the consumer does not stop after a timeout.

### To reproduce
```
from datetime import timedelta
import json
import os
from time import sleep
from threading import Thread

from hop import Stream
from hop.auth import Auth
from hop.io import StartPosition

import logging
logging.basicConfig(format='%(levelname)s:%(message)s', level=logging.INFO)


class MyListener(Stream):

    def __init__(self, *args, **kwargs):
        auth = Auth(os.getenv("HOP_USERNAME"), os.getenv("HOP_PASSWORD"))
        super().__init__(auth=auth)
        self.batch_size = 1
        self.until_eos = True
        self.running = False
        self.start_at = StartPosition.LATEST

    def listen(self, topic, callback=None):
        self.running = True
        s = self.open('kafka://kafka.scimma.org:9092/'+topic, 'r')
        while self.running:
            print("### Another loop ###")
            try:
                try:
                    for msg in s.read(batch_size=1, batch_timeout=timedelta(seconds=2)):
                        try:
                            msg = json.loads(msg.content)
                            print(msg['uid'], msg['alert_type'])
                        except AttributeError:
                            print(msg)
                except KeyError:
                    pass
            except (KeyboardInterrupt, SystemExit):
                self.running = False
        s.close()



alert_client = MyListener()

t = Thread(target=alert_client.listen, args=('gracedb-test.mdc_superevent',))
print("Starting thread")
t.start()
sleep(10)  # during this time, no messages are sent to the topic
print("Setting running to False")
alert_client.running = False
print("Calling join")
t.join()
print("Thread is_alive:", t.is_alive())
```

### Output:
```
$ python run_listener_in_thread.py
Starting thread
INFO:group ID not specified, generating a random group ID: deep.chatterjee-31ddb6ac-2IQI1T2IF3
INFO:connecting to kafka://kafka.scimma.org:9092
INFO:subscribing to topics: ['gracedb-test.mdc_superevent']
### Another loop ###
INFO:processing messages from stream
INFO:finished processing messages
### Another loop ###
INFO:processing messages from stream
Setting running to False
Calling join
< script never terminates >
```

### Possible solution
Explicitly terminating consumer runloop in the event of no messages. Diff: https://github.com/astronomy-commons/adc-streaming/compare/master...deepchatterjeeligo:adc-streaming:empty-messages

```
$ python run_listener_in_thread.py
Starting thread
INFO:group ID not specified, generating a random group ID: deep.chatterjee-31ddb6ac-1QLXKUZ6DE
INFO:connecting to kafka://kafka.scimma.org:9092
INFO:subscribing to topics: ['gracedb-test.mdc_superevent']
### Another loop ###
INFO:processing messages from stream
INFO:finished processing messages
### Another loop ###
INFO:processing messages from stream
INFO:finished processing messages
### Another loop ###
INFO:processing messages from stream
INFO:finished processing messages
### Another loop ###
INFO:processing messages from stream
INFO:finished processing messages
### Another loop ###
INFO:processing messages from stream
INFO:finished processing messages
### Another loop ###
INFO:processing messages from stream
Setting running to False
Calling join
INFO:finished processing messages
INFO:closing connection
Thread is_alive: False
```

**Comment by myNameIsPatrick** [comment]
@deepchatterjeeligo, your use case works for me regardless of what `until_eos` is set to if I modify where the `self.running` check is done:

```python
      def listen(self, topic, callback=None):
          self.running = True
          s = self.open('kafka://kafka.scimma.org:9092/'+topic, 'r')
          print("### Another loop ###")
          try:
              try:
                  for msg in s.read(batch_size=1, batch_timeout=timedelta(seconds=2)):
                      try:
                          msg = json.loads(msg.content)
                          print(msg['uid'], msg['alert_type'])
                      except (AttributeError, TypeError):
                          print(msg)
                      if not self.running:
                          break
              except KeyError:
                  pass
          except (KeyboardInterrupt, SystemExit):
              self.running = False
          s.close()
```

This is still an issue in your case, however, in that `s.read()` needs to return at least one message since `self.running` was set to False, which may not happen in a timely manner or ever.

> Issue: If there is no message in the topic, the consumer does not stop after a timeout.

I don't think this is the issue though. The timeout is set up for each iteration internally in read (`consumer.consume()`), but `s.read()` blocks indefinitely unless `until_eos=True` and the EOS condition is met (if `until_eos=True`).

Correct me if I'm wrong, but I think what you want is a way to tell hop-client/adc-streaming to stop listening to messages out-of-band when `s.read()` is blocking. If that's the case, then a `stop()` (or similar) method could be added that allows one to break out of the loop within the Consumer in these two places:

* https://github.com/astronomy-commons/adc-streaming/blob/master/adc/consumer.py#L124
* https://github.com/astronomy-commons/adc-streaming/blob/master/adc/consumer.py#L156

**Comment by cnweaver** [comment]
> Correct me if I'm wrong, but I think what you want is a way to tell hop-client/adc-streaming to stop listening to messages out-of-band when s.read() is blocking. 

I think that is the core concern, yes. 

> If that's the case, then a stop() (or similar) method could be added that allows one to break out of the loop within the Consumer in these two places:

That sounds like a good approach to me; I will have to experiment with it a bit.

**Comment by deepchatterjeeligo** [comment]
> This is still an issue in your case, however, in that s.read() needs to return at least one message since self.running was set to False, which may not happen in a timely manner or ever.

Yes. That is the main point I wanted to make. It is definitely an issue for applications that have the listener running in a different thread since it will not terminate until there is one message in the topics being listened to. We are facing this with gwcelery now where condor_rm does not stop the jobs, and we are having to do a kill -9.

> If that's the case, then a stop() (or similar) method could be added that allows one to break out of the loop within the Consumer in these two places:

Exactly.

**Comment by deepchatterjeeligo** [comment]
Hello @cnweaver and @myNameIsPatrick , I was testing out a branch for adc-streaming with a `running` attribute and a `stop` method as a part of gwcelery, and it is working fine for gwcelery, at least for a few days of testing. I created a merge request. The change should be backward compatible. This will allow us to stop the listener in gwcelery. But I'm happy to take feedback.

**Comment by deepchatterjeeligo** [comment]
this was fixed in [v2.2](https://github.com/astronomy-commons/adc-streaming/releases/tag/v2.2.0). Closing issue.
